### ISSUE TITLE
Auto dependency does not include all  files

### ISSUE THREAD

**Issue by Poolshark** [szaghi/FoBiS#108]
Hi!
First of all, thanks for the great work!
Unfortunately I can't get my build to work. I have the following (horrible) code structure - it's a scientific code and like most of them the structure is quite bad. Basically there is a **core module** and a **branch module** where each of those contains some individual `.f90` files with modules. The main program lies in the core but for execution we need e.g. a **config module** which lies in the branch folder. The `fobos.make` also is in the branch folder:
```text
PROGRAM
   L core
        |- src
             |- program.f90
             |- someModule.f90
             |- ...
             |- sub1
                  |- moreModules.f90
                  |- ...
             |- sub2
             |- ...
   L branch
        |- src
             |- configModule.f90
             |- sub1
                  |- moreModules.f90
                  |- ...
             |- sub2
             |- ...
        |- fobos.make
        |- ...
```
I think you get the idea. The problem now is that when I compile I get the error that e.g.
`Attention: the file '../core/program.f90' depends on 'config' that is unreachable`. The sample `fobos.make` looks like this:
```text
[default]
help      = This is the help message...
colors    = True
quiet     = False
jobs      = 1
compiler  = gnu
mpi       = False
cflags    = -ffpe-trap=invalid,zero,overflow -ffree-line-length-none -fintrinsic-modules-path ./build -J./build -cpp
include   = ./src ../core/sub1 ../core/sub2
extensions = .f90
mod_dir   = ./mod/
obj_dir   = ./obj/
build_dir = ./build/
src       = ../core/src/
log       = True
target    = program.f90
output    = FoO
```
Could you please help me with this?
Thanks!

**Comment by szaghi** [comment]
@Poolshark 

Hi, I think I can help you (indicating the correct use or patching the searching algorithm if it is not such a comprehensive as I had erroneously assumed), but let me few hours, maybe later today I'll have a break that I can use for checking this scenario, I am not sure I had considered it previously.

**Comment by szaghi** [comment]
@Poolshark 

I found the time for a little test... here we are: you should be able to compile your project simply adding the multiple **src* directories to the main `src` option, i.e. in you fobos you should put something along the line:

```ini
...
src = ../core/src/ ../branch/src
...
```

My test is the following:

```bash
┌╼ stefano@zaghi(04:38 PM Mon Mar 04)
├───╼ ~/downloads/fobis_test 2 files, 136Kb
└──────╼ tree
.
├── branch
│   └── src
│       └── library.f90
└── core
    └── src
        └── main.f90

4 directories, 2 files

cd core
┌╼ stefano@zaghi(04:38 PM Mon Mar 04)
├───╼ ~/downloads/fobis_test/core 1 files, 12Kb
└──────╼ FoBiS.py build -t src/main.f90 --src ../core/src/ ../branch/src/
Builder options
  Directories
    Building directory: "."
    Compiled-objects .o   directory: "obj"
    Compiled-objects .mod directory: "mod"
  Compiler options
    Vendor: "gnu"
    Compiler command: "gfortran"
    Module directory switch: "-J"
    Compiling flags: "-c"
    Linking flags: ""
    Preprocessing flags: ""
    Coverage: False
    Profile: False
  Preprocessor used: None
  Preprocessor output directory: None
  Preprocessor extensions processed: []

Building ../core/src/main.f90
Compiling ../branch/src/library.f90 serially
Compiling ../core/src/main.f90 serially
Linking ./main
Target ../core/src/main.f90 has been successfully built

cd ..
┌╼ stefano@zaghi(04:38 PM Mon Mar 04)
├───╼ ~/downloads/fobis_test 2 files, 136Kb
└──────╼ tree
.
├── branch
│   └── src
│       └── library.f90
└── core
    ├── main
    ├── mod
    │   └── library.mod
    ├── obj
    │   ├── library.o
    │   └── main.o
    └── src
        └── main.f90

6 directories, 6 files
```

Let me know if this works for your actual scenario.

**Comment by Poolshark** [comment]
Thanks a lot! I will try it out tomorrow!

**Comment by Poolshark** [comment]
Sorry, not sure if I was supposed to just update my previous post. Not sure if you got notified by my update and don't want to cause the issue as unresolved.

**Comment by szaghi** [comment]
Hi @Poolshark , do not worry, indeed I had missed the update, my bad.

So, you can execute shell commands in fobos defining  `rules`, see https://github.com/szaghi/FoBiS/wiki/fobos-rules but I am not sure this is what you want to do.

For *non module* files I can help (I hope), you can either exclude files/paths or FoBiS should be able to compile and link non module files. Let me know more details.

**Comment by Poolshark** [comment]
Hi @szaghi !

Thanks for all the inputs. I've now totally restructured the code and therefore I'm not sure if I have done things correctly so far. I think I definitely want to run shell commands as I have written a short script which sets up an environment.

What I've noticed is that even when turning on the *verbose* mode I don't get specific compile errors but something like "there might be an error in the module". This makes it super hard to develop as I have to pull out parts of the code and compile those junks by hand to test if it's working. Is there any chance to dump the whole compile output into the log file?

Thanks!

**Comment by szaghi** [comment]
Hi @Poolshark 

> What I've noticed is that even when turning on the verbose mode I don't get specific compile errors but something like "there might be an error in the module". This makes it super hard to develop as I have to pull out parts of the code and compile those junks by hand to test if it's working. Is there any chance to dump the whole compile output into the log file?

in verbose mode FoBiS simply echo all the output of the compiler, thus if the information into the log is not enough I think that it is not related to FoBiS, rather to the compiling flags. How do you compile? Do you use a "bullet prof" debugging flags? In general, the classical ´-g -C´ are not enough for a comprehensive log.

**Comment by Poolshark** [comment]
OK! That is weird because I've used the same compile flags in my *makefile* and it did show the errors in the console. This is the compile flags I'm using with FoBiS (compiling with GNU Fortran):

> `cflags    = -ffpe-trap=invalid,zero,overflow -ffree-line-length-none -cpp -g -fbacktrace -fcheck=all`

The logfile shows:

> `f951: Fatal Error: Module file ‘contbase.smod’ has not been generated, either because the module does not contain a MODULE PROCEDURE or there is an error in the module.
compilation terminated.`

and pull out the *base module* and the corresponding *sub modules*, put them into a file and compile without flags I get:

> `Error: Expected label ‘equ’ for END SUBROUTINE statement at (1)`

As there was a label conflict with the subroutine.

Do you have any suggestions for compile flags in development mode which might solve this?

**Comment by szaghi** [comment]
@Poolshark 

Submodules should be supported, but they are not well tested because I have not yet used them extensively. Can you create a Minimal Working Example reproducing the bug?

**Comment by Poolshark** [comment]
@szaghi 

I've attached a mini working example. The example consist of 3 parts:

1. A FoBiS makefile and a typical folder structure within my code - fails
2. A FoBiS makefile and all the source is contained in one `src` folder - fails
3. A single folder with all the source for manual compilation. I've included a `manual_compile.txt` with the single compile steps I've used. - working

Thanks for helping me out with this! 

[FoBiS Test.tar.gz](https://github.com/szaghi/FoBiS/files/2963253/FoBiS.Test.tar.gz)

**Comment by szaghi** [comment]
Hi @Poolshark 

sorry for my delay, it has been a busy day.

In your `fobos.make` files you have missed the `-c` flag into your `cflags` option definitions, add the missing `-c` and it should work in both your scenario, at least it worked for me.

Cheers

**Comment by Poolshark** [comment]
Hey @szaghi 

thanks, indeed the example works fine but unfortunately applying it to the main project does not work. I think I was able to pin point the problem - the **ContBase** object does not get generated (can't see it in the build folder) even though everything looks fine on the console output:

```text
Executing: gfortran -c -g -ffpe-trap=invalid,zero,overflow -ffree-line-length-none -fbacktrace -fcheck=all -J build/mod -Ibuild/mod src/physics/eqCont/ContBase.f90 -o build/obj/ContBase.o
```
At first I thought that there might be a problem because  *ContBase* has more dependencies but I've tried that in the mini example and it did work.

Any idea what could possibly go wrong here?
